Lexing: add set_position and set_filename (same as in Sedlexing)

ksromanov opened the pull request:
When we use `Parse` module functions in PPXes, sometimes it is very convenient to shift the initial location from default "line 1; column 1". For instance, see https://github.com/janestreet/ppx_custom_printf/blob/master/src/ppx_custom_printf.ml#L163
Otherwise PPX author needs to traverse AST and manually shift location in every node - it is very risky from compatibility point of view and complicated.

In `Sedlex` authors added functions `set_position` and `set_filename`, which provide such an API - https://github.com/ocaml-community/sedlex/blob/master/src/lib/sedlexing.ml#L82. 

This PR adds similar API to Stdlib `Lexing` module. Of course, we can always alter internal fields of `lexbuf`, but it is better to have compatibility between two lexers and some API.

Also, I found by experiment and checked by reading the code, that to shift location we need not only change `lexbuf.lex_curr_p`, but also `lexbuf.lex_abs_pos`.

ksromanov committed:
Add set_position and set_filename to Lexing.

gasche commented on stdlib/lexing.mli:
```diff
@@ -108,6 +108,15 @@ val from_function : ?with_positions:bool -> (bytes -> int -> int) -> lexbuf
+val set_position : lexbuf -> position -> unit
+(** Set the initial tracked input position for [lexbuf] to custom value.
+   Does not set file (use [set_file] for it). Can be used to adjust location
+   information without need to traverse AST. *)
```
```
@since 4.10
```

dra27 replied:
More comments on the doc comment:

- "to **a** custom value" or just "Set the initial tracked input position for a [lexbuf]."
- If the API stands, it should be `{!set_filename}`, not `[set_file]`, although I'd change the whole sentence to something like "Ignores [pos_fname], see {!set_filename} for changing this field."
- The last sentence is not relevant to `Lexing` and I don't think should be here - `Lexing` has nothing to do with ASTs.

gasche commented on stdlib/lexing.mli:
```diff
@@ -108,6 +108,15 @@ val from_function : ?with_positions:bool -> (bytes -> int -> int) -> lexbuf
+
+val set_filename: lexbuf -> string -> unit
+(** [set_filename lexbuf file] sets the filename to [file] in
+    [lexbuf]. *)
```
```
@since 4.10
```

dra27 replied:
The comment is misleading if the function is (erroneously) called after lexing has started, since it only updates `lex_curr_p`. Given that this function is supposed to be called *before* lexing begins, wouldn't it be better embedded as an optional argument in the `from_` functions?

ksromanov replied:
> The comment is misleading if the function is (erroneously) called after lexing has started, since it only updates lex_curr_p.

Thanks! We can update `lex_start_p` as well. It is very unlikely that this overhead creates performance problems.

> wouldn't it be better embedded as an optional argument in the from_ functions?

1. I wanted to have `Lexing` and `Sedlexing` be as compatible as possible.
2. If I understand it right, both functions can be called during lexing - this will shift the positions of next tokens. To make this more reliable, we can alter both `lex_start_p` and `lex_curr_p`.

Does this plan (implement two proposed functions, shift both `lex_start_p` and `lex_curr_p`) look good for you?

gasche replied:
For the record, I'm not fond of adding optional arguments to `from_*` functions. I think that APIs resulting from this process and heavyweight and not-so-pleasant to use.

I would prefer doing `with_pos p (Lexing.from_string s)` over `Lexing.from_string ~pos:p s`. Given that lexing buffers are mutable, a `set_position`  mutator feels reasonable and consistent to me.

gasche reviewed (commented):
I agree with the use-case (starting parsing something that is at a known location in a string literal in an AST) and I think the API proposed is good. The fact that, if I understand properly, people were previously not setting the current position correction (by not updated `lex_abs_pos` is a strong case in favor of higher-level functions.

I would like to see tests to ensure that the planned use-case works as desired. (Call a small lexer several times, and then create a new buffer to restart lexing from that point, and print the resulting locations?) Do you have an example that behaves incorrectly when `lex_abs_pos` is not set? What does it do in that case?

dra27 commented on stdlib/lexing.mli:
```diff
@@ -108,6 +108,15 @@ val from_function : ?with_positions:bool -> (bytes -> int -> int) -> lexbuf
    starting at index 0, and return the number of bytes
    provided. A return value of 0 means end of input. *)
 
+val set_position : lexbuf -> position -> unit
```
I'm slightly uneasy at accepting a record where a field is ignored, but I'm also not particularly familiar with doing it. Why is

```OCaml
val set_position : lexbuf -> int -> int -> int -> unit
```
not more appropriate? (with appropriate discussion on labelling, or using a triple)?

ksromanov replied:
I wanted this function be compatible with

`val set_position: lexbuf -> Lexing.position -> unit` in [Sedlexing](https://github.com/ocaml-community/sedlex/blob/master/src/lib/sedlexing.mli#L51)

damiendoligez replied:
I think it's OK to ignore one field of the record. At any rate, better than passing a triple or three unlabeled arguments.

ksromanov commented:
Thank you very much for comments!

@gasche - here is an example:

```ocaml
let position = Lexing.{ (* This corresponds to File "file.ml", line 100, character 10 *)
    pos_fname = "file.ml";
    pos_lnum = 100;
    pos_bol = 1000;
    pos_cnum = 1010;
}

(* Good version *)
let set_position lexbuf position =
  let open Lexing in
  lexbuf.lex_curr_p  <- {position with pos_fname = lexbuf.lex_curr_p.pos_fname};
  lexbuf.lex_abs_pos <- position.pos_cnum

(* Incomplete version *)
let set_position_incomplete lexbuf position =
  let open Lexing in
  lexbuf.lex_curr_p  <- {position with pos_fname = lexbuf.lex_curr_p.pos_fname}

(* "Testing framework" *)
let print_error_in_parse set_position_variant = 
    try
        let _ = 
            let lexbuf = Lexing.from_string "f ( x" in (* contains error in char 3, counting from 1 *)
            set_position_variant lexbuf position;
            Parse.expression lexbuf in ()
    with e -> Location.report_exception Format.std_formatter e

let _ = 
    print_string "Incomplete version:\n";
    print_error_in_parse set_position_incomplete;
    print_string "Good version:\n";
    print_error_in_parse set_position
```

output (compiler 4.08):

```bash
kromanov:~/test$ ocamlc -I +compiler-libs ocamlcommon.cma  test.ml 
kromanov:~/test$ ./a.out 
Incomplete version:
File "_none_", line 100, characters -995--995:
Error: Syntax error: ')' expected
File "_none_", line 100, characters -998--997:
  This '(' might be unmatched
Good version:
File "_none_", line 100, characters 15-15:
Error: Syntax error: ')' expected
File "_none_", line 100, characters 12-13:
  This '(' might be unmatched
```

_Complete version_ may be gives result, which is off by 1, but _incomplete version_ produces complete garbage. This is because the final `pos_cnum` is produced by 
`buf.lex_curr_p <- {buf.lex_curr_p
                       with pos_cnum = buf.lex_abs_pos + buf.lex_curr_pos};`
(see https://github.com/ocaml/ocaml/blob/trunk/stdlib/lexing.ml#L69 for old engine). So it completely overrides the value in `lex_curr_p` using `lex_abs_pos`.

nojb commented:
If the objective is to run a lexer from some position other than the initial one, shouldn't `lex_curr_pos` be updated as well? Am I missing something?

ksromanov commented:
> If the objective is to run a lexer from some position other than the initial one, shouldn't `lex_curr_pos` be updated as well? Am I missing something?

The objective is to fake location. So, that in the result of `Parse.expression @@ Lexing.from_string "1 + 2"` all nodes would have coherently shifted locations. In the example above, it is like `file.ml, line 100, column 10`.

`lex_curr_pos` is the _real_ location inside the string. While `lex_abs_pos` is an offset of the beginning of the string. So, we need to shift only `lex_abs_pos`.

nojb commented:
> The objective is to fake location. So, that in the result of `Parse.expression @@ Lexing.from_string "1 + 2"` all nodes would have coherently shifted locations. In the example above, it is like `file.ml, line 100, column 10`.

OK, thanks.

gasche commented:
@ksromanov thanks for the test. Could you include it in the compiler testsuite, for example as a new file `testsuite/tests/parsing/change_start_loc.ml`?

ksromanov commented:
@gasche 
> Could you include it in the compiler testsuite, for example as a new file `testsuite/tests/parsing/change_start_loc.ml`?

Thanks for giving the path - I am totally unfamiliar with testsuite. :-(

@dra27 , @gasche 

1. I've added test, which checks both functions.

2. Changed comments.

3. I decided, that it is ok in `set_position` to alter only `lexbuf.lex_curr_p` and leave `lexbuf.lex_start_p` as it is, because `lexbuf.lex_curr_p` indicates starting position for the next token. On the other hand changing `lexbuf.lex_start_p` would be retroactive - it is starting position of already fetched token. So, if the user wants to call `set_position` and `set_filename` in the middle of lexing, he should be quite careful.

Does it look good for you?

gasche reviewed (approved):
I approve of the change. I won't merge now, because I would like to give @dra27 some time to comment and request additional changes if he feels strongly about them.

gasche commented:
And: you need a Changes entry.

pmetzger commented on Changes:
```diff
@@ -53,6 +53,10 @@ Working version
    (Jacques-Henri Jourdan, review by Stephen Dolan and Gabriel Scherer)
 
 ### Standard library:
+- #8771: Lexing: add set_position and set_filename to shift (fake)
```
The word "shift" is a bit confusing here (at least to me). "set" isn't accurate?

pmetzger replied:
(Or perhaps "to change the initial tracking position of the lexbuf"?)

ksromanov replied:
Thanks for review, fixed!

Are the comments in lexing.mli clear? There should be no confusion, that we are not starting actual tokenization from a specific position of the file/string, but we essentially _fake_ the position, which will be inserted in to parse errors/AST nodes.

pmetzger replied:
Everything does seem clear enough now, yes.

shindere commented:
Thanks a lot for having rebased and squashed.

At the moment your test only compiles the program and checks the compiler's
output -- the program is not actually run. To make sure the program gets
run, you can replace the three actions you have used (the lines starting
with a star) by just the "bytecode" test. You could also specify no
test at all and then ocamltest would run the default tests.
Also, you don't need to indent the lines in the test block, I think.

Could you please update your commit accordingly? No need to do a separate
commit, just force-push an updated version of the same one, please.

ksromanov commented:
> Could you please update your commit accordingly? No need to do a separate
> commit, just force-push an updated version of the same one, please.

Thanks a lot for working on this PR. It is updated now:

1. Added you as a reviewer to Changes along with Gabriel and David.

2. Removed actions from test code - thanks a lot for pointing this out.
I've also checked that it indeed runs the code (I've broken test - it failed)

3. Changed comments and position (from line 100: col 0 to line 100: col 10)
in TEST code (main code is UNTOUCHED!).

4. Moved Changes entry from 4.10 to working version.

5. Squashed again.

gasche commented on Changes:
```diff
@@ -312,7 +317,6 @@ OCaml 4.10.0
   (Gabriel Scherer, review by Florian Angeletti)
 
 ### Standard library:
-
```
Minor nitpick: this line removal shouldn't be in the patch. (But it's fine if we merge with it, I can fix the Changes afterwards.)

ksromanov replied:
Fixed. Thanks for noticing.

gasche reviewed (commented):
Thanks for the updates. This looks good to go when the CI comes back green.

shindere commented:
As I said, the lines in the TEST block do not really need to be
indented so if you don't mind updating the commit just with that it'd be
great.

ksromanov commented:
Fixed.

gasche commented:
@jhjourdan AppVeyor mingw32 fails on this PR on a statmemprof test, `lists_in_minor.ml`:

```diff
check_callstack
+Called from unknown location
Raised by primitive operation at file "lists_in_minor.ml", line 14, characters 11-33
Called from file "lists_in_minor.ml", line 69, characters 2-26
Called from file "lists_in_minor.ml", line 76, characters 2-20
OK !
```

I'm confident this is unrelated to the present PR, so I will merge now.

gasche commented:
@shindere about half of the tests in the testsuite have an indented TEST block. Two reasons to do it is that it's rather natural to highlight the block structure (of course people may also use syntactic highlighting to do this), and that this is the style that Tuareg will automatically use for indentation. I don't think we should ask people to change their indentation on each PR.

shindere commented:
Konstantin Romanov (2019/11/16 14:18 -0800):
> Fixed.

Thanks!
